Sources/PenguinParallel/ThreadPool.swift

```diff
@@ -91,105 +127,31 @@ public protocol ComputeThreadPool {
-      try withUnsafePointer(to: &holder) { holder in
-        try runParallelFor(pool: self, start: 0, end: n, total: n, fn: holder)
+  /// Convert a non-vectorized operation to a vectorized operation.
+  public func parallelFor(n: Int, _ fn: ParallelForFunction) {
+    parallelFor(n: n) { start, end, total in
+      for i in start..<end {
+        fn(i, total)
       }
     }
   }
-}
 
-/// Typed compute threadpools support additional sophisticated operations.
-public protocol TypedComputeThreadPool: ComputeThreadPool {
-  /// Submit a task to be executed on the threadpool.
-  ///
-  /// `pRun` will execute task in parallel on the threadpool and it will complete at a future time.
-  /// `pRun` returns immediately.
-  func dispatch(_ task: (Self) -> Void)
-
-  /// Run two tasks (optionally) in parallel.
-  ///
-  /// Fork-join parallelism allows for efficient work-stealing parallelism. The two non-escaping
-  /// functions will have finished executing before `pJoin` returns. The first function will execute on
-  /// the local thread immediately, and the second function will execute on another thread if resources
-  /// are available, or on the local thread if there are not available other resources.
-  func join(_ a: (Self) -> Void, _ b: (Self) -> Void)
-
-  /// Run two throwing tasks (optionally) in parallel; if one task throws, it is unspecified
-  /// whether the second task is even started.
-  ///
-  /// This is the throwing overloaded variation.
-  func join(_ a: (Self) throws -> Void, _ b: (Self) throws -> Void) throws
-}
-
-extension TypedComputeThreadPool {
-  /// Implement the non-throwing variation in terms of the throwing one.
-  public func join(_ a: (Self) -> Void, _ b: (Self) -> Void) {
-    withoutActuallyEscaping(a) { a in
-      let throwing: (Self) throws -> Void = a
-      // Implement the non-throwing in terms of the throwing implementation.
-      try! join(throwing, b)
+  /// Convert a non-vectorized operation to a vectorized operation.
+  public func parallelFor(n: Int, _ fn: ThrowingParallelForFunction) throws {
+    try parallelFor(n: n) { start, end, total in
+      for i in start..<end {
+        try fn(i, total)
+      }
     }
   }
 }
 
-extension TypedComputeThreadPool {
-  public func dispatch(_ fn: @escaping () -> Void) {
-    dispatch { _ in fn() }
-  }
-
-  public func join(_ a: () -> Void, _ b: () -> Void) {
-    join({ _ in a() }, { _ in b() })
-  }
-
-  public func join(_ a: () throws -> Void, _ b: () throws -> Void) throws {
-    try join({ _ in try a() }, { _ in try b() })
-  }
-}
-
 /// A `ComputeThreadPool` that executes everything immediately on the current thread.
 ///
 /// This threadpool implementation is useful for testing correctness, as well as avoiding context
 /// switches when a computation is designed to be parallelized at a coarser level.
-public struct InlineComputeThreadPool: TypedComputeThreadPool {
+public struct InlineComputeThreadPool: ComputeThreadPool {
```

Comment:
```suggestion
public struct SerialExecution: ParallelizableExecutionStrategy {
```

Consider the protocol name an initial suggestion.  `ComputeThreadPool` seems wrong to me, as models are not necessarily thread pools.

When we do mention `Thread`s, is there any point in adding `Compute`?

Reply:
> In terms of thread-pools, there can be a number of different designs with different properties. 

Naturally.

> In the same way that you can implement a random access collection in terms of a collection (just really inefficiently), 

You mean like [`Sampling<SomeCollection, [SomeCollection.Index]>`](https://github.com/tensorflow/swift-apis/pull/967/files#r429669380)?  (Phew, glad I looked at that just now!)

> I wanted to clearly distinguish what properties the thread-pool has. 

Also naturally.  I don't get the connection to random access, though.

> Concretely, there are I/O-focused thread-pools, where you can blocking and/or non-blocking I/O. This thread pool abstraction is focused on compute-bound tasks, and is tuned / structured with APIs focused on that domain. Does that make sense?

In principle.  Are they really separate abstractions though, or at least, isn't the IO one a refinement of this one?  Wouldn't you want to write most algorithms once rather than replicate them for different kinds of pools?